gdal/frmts/wcs/wcsutils.cpp

```diff
@@ -0,0 +1,698 @@
+                cache_dir = CPLFormFilename(dir, subdir, NULL);
+            }
+        }
+    }
+    cache_dir = CPLFormFilename(cache_dir, "wcs_cache", NULL);
+    if (!MakeDir(cache_dir)) {
+        return false;
+    }
+    if (clear) {
+        char **folder = VSIReadDir(cache_dir);
+        int size = folder ? CSLCount(folder) : 0;
+        for (int i = 0; i < size; i++) {
+            if (*folder[i] == '.') {
+                continue;
+            }
+            CPLString filepath = CPLFormFilename(cache_dir, folder[i], NULL);
+            remove(filepath);
+        }
+        CSLDestroy(folder);
+    }
+    return true;
+}
+
+/* -------------------------------------------------------------------- */
+/*      DeleteEntryFromCache                                            */
+/* -------------------------------------------------------------------- */
+
+bool DeleteEntryFromCache(CPLString cache_dir, CPLString  key, CPLString value)
+{
+    // depending on which one of key & value is not "" delete the relevant entry
+    CPLString db_name = CPLFormFilename(cache_dir, "db", NULL);
+    char **data = CSLLoad(db_name); // returns NULL in error and for empty files
+    char **data2 = CSLAddNameValue(NULL, "foo", "bar");
+    CPLString filename = "";
+    if (data) {
+        for (int i = 0; data[i]; ++i) {
+            char *val = strchr(data[i], '=');
+            if (*val == '=') {
+                *val = '\0';
+                val = val + 1;
+                if ((key != "" && key == data[i])
+                    || (value != "" && value == val)
+                    || (strcmp(data[i], "foo") == 0))
+                {
+                    if (key != "") {
+                        filename = data[i];
+                    } else if (value != "") {
+                        filename = data[i];
+                    }
+                    continue;
+                }
+                data2 = CSLAddNameValue(data2, data[i], val);
+            }
+        }
+        CSLDestroy(data);
+    }
+    CSLSave(data2, db_name); // returns 0 in error and for empty arrays
+    CSLDestroy(data2);
+    if (filename != "") {
+        char **folder = VSIReadDir(cache_dir);
+        int size = folder ? CSLCount(folder) : 0;
+        for (int i = 0; i < size; i++) {
+            if (*folder[i] == '.') {
+                continue;
+            }
+            CPLString name = folder[i];
+            if (name.find(filename) != std::string::npos) {
+                CPLString filepath = CPLFormFilename(cache_dir, name, NULL);
+                remove(filepath);
+            }
+        }
+        CSLDestroy(folder);
+    }
+    return true;
+}
+
+/* -------------------------------------------------------------------- */
+/*      FromCache                                                       */
+/* -------------------------------------------------------------------- */
+
+int FromCache(CPLString cache_dir, CPLString &filename, CPLString url)
```

Comment:
const CPLString& cache_dir and url

So retval can be 0, 1 or -1 ? Perhaps use an enum instead of int for the return value